google/cloud/storage/internal/hash_validator_test.cc

```diff
@@ -138,74 +181,73 @@ TEST(Crc32cHashValidator, Simple) {
-    EXPECT_EQ("ImIEBA==", ex.computed_hash());
-    throw;
-  }, std::ios::failure);
-#else
-  auto result = std::move(validator).Finish("test msg");
+  auto result = std::move(validator).Finish();
   EXPECT_EQ("<invalid-value-for-test>", result.received);
   EXPECT_EQ("ImIEBA==", result.computed);
-#endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 }
 
 TEST(Crc32cHashValidator, MultipleHashesCrc32cAtEnd) {
   Crc32cHashValidator validator;
   validator.Update("The quick");
   validator.Update(" brown");
   validator.Update(" fox jumps over the lazy dog");
-  validator.ProcessHeader(
-      "x-goog-hash", "md5=<ignored>,crc32c=<invalid-value-for-test>");
+  validator.ProcessHeader("x-goog-hash",
+                          "md5=<ignored>,crc32c=<invalid-value-for-test>");
   // I used this command to get the expected value:
   // /bin/echo -n "The quick brown fox jumps over the lazy dog" > foo.txt
   // gsutil hash foo.txt
-#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
-  EXPECT_THROW(try {
-    std::move(validator).Finish("test msg");
-  } catch(google::cloud::storage::HashMismatchError const& ex) {
-    EXPECT_EQ("<invalid-value-for-test>", ex.received_hash());
-    EXPECT_EQ("ImIEBA==", ex.computed_hash());
-    throw;
-  }, std::ios::failure);
-#else
-  auto result = std::move(validator).Finish("test msg");
+  auto result = std::move(validator).Finish();
   EXPECT_EQ("<invalid-value-for-test>", result.received);
   EXPECT_EQ("ImIEBA==", result.computed);
-#endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 }
 
 TEST(Crc32cHashValidator, MultipleHashes) {
   Crc32cHashValidator validator;
   validator.Update("The quick");
   validator.Update(" brown");
   validator.Update(" fox jumps over the lazy dog");
-  validator.ProcessHeader(
-      "x-goog-hash", "crc32c=<invalid-value-for-test>,md5=<ignored>");
+  validator.ProcessHeader("x-goog-hash",
+                          "crc32c=<invalid-value-for-test>,md5=<ignored>");
   // I used this command to get the expected value:
   // /bin/echo -n "The quick brown fox jumps over the lazy dog" > foo.txt
   // gsutil hash foo.txt
-#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
-  EXPECT_THROW(try {
-    std::move(validator).Finish("test msg");
-  } catch(google::cloud::storage::HashMismatchError const& ex) {
-    EXPECT_EQ("<invalid-value-for-test>", ex.received_hash());
-    EXPECT_EQ("ImIEBA==", ex.computed_hash());
-    throw;
-  }, std::ios::failure);
-#else
-  auto result = std::move(validator).Finish("test msg");
+  auto result = std::move(validator).Finish();
   EXPECT_EQ("<invalid-value-for-test>", result.received);
   EXPECT_EQ("ImIEBA==", result.computed);
-#endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+}
+
+TEST(CompositeHashValidator, Empty) {
+  CompositeValidator validator(
+      google::cloud::internal::make_unique<Crc32cHashValidator>(),
+      google::cloud::internal::make_unique<MD5HashValidator>());
+  // Look at the tests for each validator to find out where the magic numbers
```

Comment:
If we're reusing these in multiple places, it's probably worth storing them as constants that are shared across all the places they're used.  E.g. this could be `std::string const B64_MD5_FOR_EMPTY_STR = 1B2M2Y8AsgTpgAmY7PhCfg==`, which is self explanatory and doesn't require instructions to go hunt down an explanation :)

Reply:
Great idea, thanks!